webapp/src/lib/Ui.ts

```diff
@@ -73,6 +102,121 @@ export interface IDialogConfig {
 export class DialogOKCancelComponent {
     constructor(
         public dialogRef: MatDialogRef<DialogOKCancelComponent>,
-        @Inject(MAT_DIALOG_DATA) public data: IDialogConfig) {
+        @Inject(MAT_DIALOG_DATA) public data: IDialogOKCConfig) {
+    }
+}
+
+export interface IDialogTransactionConfig {
+    title: string;
+    worker: TWorker;
+}
+
+@Component({
+    selector: "app-dialog-transaction",
+    styleUrls: ["./app-dialog-transaction.scss"],
+    templateUrl: "dialog-transaction.html",
+})
+export class DialogTransactionComponent implements OnInit {
+
+    static progress(dtc: DialogTransactionComponent, percentage: number, text: string) {
+        dtc.percentage = percentage >= 0 ? percentage : percentage * -1;
+
+        if (dtc.percentage >= 100) {
+            clearTimeout(dtc.ubTimer);
+            dtc.blockIndex++;
+            dtc.addBlock();
+            setTimeout(() => {
+                dtc.dialogRef.close();
+            }, 1000);
+        }
+
+        if (percentage >= 0) {
+            dtc.text = "";
+            dtc.addTransaction(text);
+        } else {
+            dtc.text = text;
+        }
+    }
+
+    percentage: number;
+    blockIndex: number;
+    blocks: Element[] = [];
+    transaction: Element;
+    error: string;
+    text: string;
+    @ViewChild("main", {static: false}) main?: ElementRef;
+    ubTimer: Timer;
+
+    constructor(
+        public dialogRef: MatDialogRef<DialogTransactionComponent>,
+        public bcv: BcviewerService,
+        private renderer: Renderer2,
+        private el: ElementRef,
+        @Inject(MAT_DIALOG_DATA) public data: IDialogTransactionConfig) {
+    }
+
+    async ngOnInit() {
+        this.bcv.updateBlocks();
+        this.ubTimer = setInterval(() => this.updateBlocks(), 500);
+        setTimeout(() => this.startTransactions(), 200);
+    }
+
+    async updateBlocks() {
+        this.bcv.updateBlocks();
+        if (this.blockIndex < this.bcv.currentBlock.index) {
+            this.blockIndex = this.bcv.currentBlock.index;
+            this.addBlock();
+        }
+    }
+
+    async startTransactions() {
+        this.blockIndex = this.bcv.currentBlock.index;
+        for (let i = -2; i <= 0; i++) {
+            this.addBlock(this.blockIndex + i);
+        }
+        const prog = (p: number, t: string) => DialogTransactionComponent.progress(this, p, t);
+        try {
+            await this.data.worker(prog);
+            prog(-100, "Done");
+        } catch (e) {
+            this.error = e.toString();
+        }
+    }
+
+    addBlock(index: number = this.blockIndex): Element {
```

Comment:
Hum, as you're doing it with a dedicated function, I think that you already tried with a template and `ngIf` magics; but, if not, that might be doable/clearer with it.

Reply:
That would be more of a `ngIf`, `ngClass` orgy. Possible, but I thought it's simpler to create the objects and add the classes here.